#QBB2022 - Day 4 - Lunch Exercises Submission

**Exercise 1**

Portion of do_all.sh output (not script) that reports how many bp each feature covers:

```
--- Subsetting exons.chr21.bed.vcf
    + Covering 1107407 bp
--- Subsetting processed_pseudogene.chr21.bed.vcf
    + Covering 956640 bp
--- Subsetting protein_coding.chr21.bed.vcf
    + Covering 13780687 bp
```

One or more strategies to confirm that reproduced plots are the same as in the cache/ directory:

* After running the code, open the plots copied from the cache/ directory and the plots generated by running the code and compare visually
* Write a script that opens and reads both png files, and tests whether they are the same using the == operator, and returns True or prints "same" if they are

Three other gene_types present in the GENCODE .gtf that you find interesting and why:

* transcribed_unprocessed\_pseudogene: these are pseudogenes that have not been moved away from their parent gene, but are still transcribed, which means maybe they are used for something. interesting because generally pseudogenes are only transcribed if they are processed?
* miRNA: non-coding regulatory RNAs. maybe interesting to compare to the parts of the genome that they interact with 
* lncRNA: always interesting because we don't know what a lot of them do

**Exercise 2**

* The protein coding, random snippet, and lncRNA allele counts have similar frequency distributions, with a peak at the lowest allele counts and a relatively uniform but decreasing distribution towards higher allele counts.
* Both the exons and processed pseudogene distributions are more segmented, each with two peaks in the higher allele count regions but otherwise clustered around lower allele counts.

**Exercise 3** 

**SYNOPSIS**
     bxlab/cmdb-plot-vcfs -- ...

**USAGE**
     bash do_all.sh <file1> <file2>

     <file1>   vcf file containing SNPs to process
	 <file2>   gtf file containing annotations of gene type

 DESCRIPTION
 
     1. Create .bed files for features of interest
         - Run subset_regions.sh Bash script
         - Use grep to subset the GTF file to only chromosome 21
		 - Use grep and awk to locate gene types protein_coding, processed_pseudogene, 
			 and lncRNA and create a bed file with their chromosome number and start/
			 stop positions (formatted to bed conventions)
		 - Use grep and awk to locate all exons and create a bed file with their 
		 	chromosome number and start/stop positions (formatted to bed conventions)
	 2.  
